core/deadline.go

```diff
@@ -49,6 +49,8 @@ type Deadliner interface {
 
 // Deadline implements the Deadliner interface.
 type Deadline struct {
+	// Consumers listen on addedChan to find out if their call to Add(duty) succeeded.
+	addedChan    chan bool
```

Comment:
This isn't going to work. You cannot guarantee which reader will receive the message if multiple are listening. Follow the. pattern in `aggsigdb.MemDB.commands.writeCommand`. but rename to:

```
type deadlineInput struct {
  Duty core.Duty
  Success chan <- bool 
}

// Note we can rename and unexport deadliner implementation since we only ever export the interface.
type deadliner struct {
  input chan deadlineInput
  ...
}
```

Reply:
Based on this i would suggest:
```
func (d *Deadline) Add(duty Duty) {
	select {
	case <-d.quit:
		return
	default:
               input := deadlineInput{duty: duty, success: make(chan bool)}
               d.input <- input
               return <-input.success
	}
}
```

Reply:
Please look closely at aggsigdb.MemDB.commands.writeCommand, the above will not work

Reply:
```
func (d *Deadline) Add(duty Duty) bool {
        response := make(chan bool)
	select {
	case <-d.quit:
		return false
	default:
               d.input <- deadlineInput{duty: duty, success: response}
	}

        select {
	case <-d.quit:
		return false
	default:
                return <-response
	}
}
```

Reply:
Question, why not used buffered channel `response := make(chan bool, 1)` like aggsigdb?

Reply:
Buffered channel would be useful to unblock deadliner's central goroutine to not wait for the reader of this channel. In case if add goroutine is exited, central go routine should also be exited not to wait for the reader to read response to add.

We should use buffered channel 👍

Reply:
I think either is actually fine in this specific use-case. "Unbuffered" handover is fine since Add goroutine will always already be listening. AggSigDB uses a buffered channel since its Store goroutine may exit early when its context is closed so might not be listening at all. In this case, `Add` doesn't have a context so it cannot exit early so buffered channel isn't required.